Use return_and_correct_aliasing() for NJT + compatible storage setting

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121947
* __->__ #126552
* #125947

Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

jbschlosser committed:
Use return_and_correct_aliasing() for NJT + compatible storage setting

[ghstack-poisoned]

bdhirsh commented on torch/csrc/Module.cpp:
```diff
@@ -2105,6 +2106,13 @@ Call this whenever a new thread is created in order to propagate values from
   py_module.def(
       "_set_neg", [](const at::Tensor& x, bool neg) { x._set_neg(neg); });
+  py_module.def(
+      "_set_storage",
```
(trying to bikeshed on the name.... `_unsafe_set_storage`/`_set_only_storage`?)

@albanD mostly wondering if you are unhappy with a dedicated API for this. Basically, `x.set_(storage, size, stride)`: does 3 things right now:

(1) set the storage
(2) call `set_sizes_strides()
(3) potentially resize the storage (if the number of bytes required for the given size/stride is larger than the nbytes of the passed-in storage

And `return_and_correct_aliasing` (used by subclasses) only ever calls `a.set_(b.storage(), a.size(), a.stride())`, where we are guaranteed that the resize and the `set_sizes_strides_` are not necessary.

Those extra two steps are (1) a pain for nested tensor (trying to compute nbytes), and (2) unnecessary work we are doing throughout tracing (since this is used by `FunctionalTensor` as well)

jbschlosser replied:
Agreed, your suggested names are much better - do we not already have something like this? Just looked and there's `_unsafe_reset_storage()` that seems to be a functionalization thing.

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch._C._set_storage()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch._C._set_storage()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch._C._set_storage()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch._C._set_storage()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch._C._set_storage()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch._C._set_storage()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser commented:
@bdhirsh this PR breaks some NJT stuff higher up in the stack. Hacking in the storage changes the reported `nt.device` to `"meta"`, which is not correct.

I think the difficulty here comes from the fact that wrapper subclasses have null storages with the desired device hacked in:
https://github.com/pytorch/pytorch/blob/980f5ac0499447cd6d39b6241ae30ae30937a3e5/torch/csrc/autograd/python_variable.cpp#L838-L846

Combine that with this scenario during tracing:
* Outer tensor has "real" storage reporting "cpu" / "cuda:0" / etc. (albeit null)
* Inner tensors are fake with meta storages

Dense -> subclass views mix the two and cause problems. Not sure if this is easily addressable for NJT; might be a better approach to dodge the debug asserts without the use of `return_and_correct_aliasing()`.

I'm happy to break out the part that does storage setting without extra check logic to help #125977.

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Use return_and_correct_aliasing() for NJT + compatible storage setting"


Fixes #125503

Context: `return_and_correct_aliasing()` is required for traceable wrapper subclasses so that aliasing relationships are correct. NJT has not been using this, but needs to for correct aliasing relationships, and to avoid tripping asserts when DEBUG=1 (e.g. #125503).

This PR:
* Uses `return_and_correct_aliasing()` in NJT
* Changes how storage setting is done in `return_and_correct_aliasing()`
    * Old way: use `set_.source_Storage_storage_offset()`, which has extra logic for storage resizing that we don't need
    * New way: `torch.ops.aten._unsafe_set_storage_()` that shoves in a storage without this extra logic. Notably, this avoids `computeStorageNbytes()` choking on nested ints in NJT's sizes / strides

[ghstack-poisoned]
